File: source/0023. Merge k Sorted Lists/main.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeKLists(self, lists: list[ListNode]) ->ListNode:
        while None in lists:
            lists.remove(None)

        if not lists: return None
        if len(lists) == 1: return lists[0]

        major = ListNode()
        dummy = major

        def minKey(ll: ListNode) -> int:
            if ll: return ll.val
            else: return 100000

        while lists:
            minIDX = lists.index(min(lists, key=minKey))
            major.next = ListNode(lists[minIDX].val)
            major = major.next
            lists[minIDX] = lists[minIDX].next
            if not lists[minIDX]: 
                lists.pop(minIDX)

        return dummy.next

# https://leetcode.com/problems/merge-k-sorted-lists/solutions/6577583/solution-for-merge-k-lists-in-python-by-p3jdg/

File: source/0023. Merge k Sorted Lists/main.cpp
#include <algorithm>
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        for (int i = 0; i < lists.size(); i++) {
            if (lists[i] == nullptr)
                lists.erase(lists.begin() + i--);
        }
        if (lists.size() == 0) 
            return nullptr;
        
        ListNode dummy = ListNode();
        ListNode* head = &dummy;
        
        while (lists.size() > 0) {
            auto it = min_element(
                lists.begin(), 
                lists.end(), 
                [] (ListNode* &a, ListNode* &b) {
                    if (a == nullptr) return false;
                    if (b == nullptr) return true;
                    return (a->val) < (b->val);
                }
            );
            ListNode* &min_list = *it;
            
            head->next = min_list;
            head = head->next;
            if (min_list->next == nullptr) lists.erase(it);
            else min_list = min_list->next;
        }
        
        return dummy.next;
    }
};